Changed filter to lfilter with zi

jdpigeon opened the pull request:
There is definitely a warmup period before things settle down, but it looks fine to me with Muse and definitely better with OpenBCI.

Try it and tell me what you think

alexandrebarachant commented:
Thanks, this is great.

There is an issue with your implementation, you filter the whole buffer every time, so filter states are not valid.
What you have to do is create a secondary data buffer, and only filter the new incoming data chunk. 
So basically, when you receive a new chunk from LSL, apply the filter and save updated filter states.

jdpigeon commented:
Think I got it working now.

I added a second, smaller data buffer that fills up with samples, is filtered and added to the plot everytime it is drawn (filt_state is saved), and then cleared to make room for new samples.

Looks good on both Muse and OpenBCI. And no flapping! :+1: 
![non-flappyeeg](https://user-images.githubusercontent.com/4932565/27504753-55ab1964-585e-11e7-8cc2-1f76306c4f33.png)

alexandrebarachant commented on lsl-viewer.py:
```diff
@@ -104,27 +104,38 @@ def __init__(self, stream, fig, axes,  window, scale, dejitter=True):
+                                                                    zi=self.filt_state)
+                        data_f = np.vstack([self.data[:-len(self.filt_buffer)], filt_samples])
+                        self.filt_buffer = np.zeros((0, self.n_chan))
+                        self.data = data_f
```
i would actually prefer to save data_f as a second buffer, so that you can switch between filtered and unfiltered data (keystroke 'd')

alexandrebarachant commented:
Looks good to me. Merged.
